src/controller/CHIPDeviceController.cpp

```diff
@@ -1209,192 +1143,167 @@ CHIP_ERROR DeviceCommissioner::ValidateAttestationInfo(const ByteSpan & attestat
-    VerifyOrExit(commissioner->mDeviceBeingCommissioned != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    err = mOperationalCredentialsDelegate->GenerateNOCChain(data.NOCSRElements, data.attestationSignature, device->GetDAC(),
+                                                            ByteSpan(), ByteSpan(), &mDeviceNOCChainCallback);
+    StopSessionIf(err);
+}
 
-    // Check if the callback returned a failure
-    VerifyOrExit(status == CHIP_NO_ERROR, err = status);
+TRAMPOLINE_4(DeviceNOCChainGeneration, CHIP_ERROR, const ByteSpan &, const ByteSpan &, const ByteSpan &);
+void DeviceCommissioner::OnDeviceNOCChainGeneration(CHIP_ERROR status, const ByteSpan & noc, const ByteSpan & icac,
+                                                    const ByteSpan & rcac)
+{
+    StopSessionIf(mState != State::Initialized, CHIP_ERROR_INCORRECT_STATE);
+    StopSessionIf(mDeviceBeingCommissioned == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    StopSessionIf(status);
 
     // TODO - Verify that the generated root cert matches with commissioner's root cert
-
-    device = commissioner->mDeviceBeingCommissioned;
+    CommissioneeDeviceProxy * device = mDeviceBeingCommissioned;
 
     {
         // Reuse NOC Cert buffer for temporary store Root Cert.
         MutableByteSpan rootCert = device->GetMutableNOCCert();
 
-        err = ConvertX509CertToChipCert(rcac, rootCert);
-        SuccessOrExit(err);
-
-        err = commissioner->SendTrustedRootCertificate(device, rootCert);
-        SuccessOrExit(err);
+        StopSessionIf(ConvertX509CertToChipCert(rcac, rootCert));
+        StopSessionIf(SendTrustedRootCertificate(device, rootCert));
     }
 
     if (!icac.empty())
     {
         MutableByteSpan icaCert = device->GetMutableICACert();
 
-        err = ConvertX509CertToChipCert(icac, icaCert);
-        SuccessOrExit(err);
-
-        err = device->SetICACertBufferSize(icaCert.size());
-        SuccessOrExit(err);
+        StopSessionIf(ConvertX509CertToChipCert(icac, icaCert));
+        StopSessionIf(device->SetICACertBufferSize(icaCert.size()));
     }
 
     {
         MutableByteSpan nocCert = device->GetMutableNOCCert();
 
-        err = ConvertX509CertToChipCert(noc, nocCert);
-        SuccessOrExit(err);
-
-        err = device->SetNOCCertBufferSize(nocCert.size());
-        SuccessOrExit(err);
-    }
-
-exit:
-    if (err != CHIP_NO_ERROR)
-    {
-        ChipLogError(Controller, "Failed in generating device's operational credentials. Error %s", ErrorStr(err));
-        commissioner->OnSessionEstablishmentError(err);
+        StopSessionIf(ConvertX509CertToChipCert(noc, nocCert));
+        StopSessionIf(device->SetNOCCertBufferSize(nocCert.size()));
     }
 }
 
-CHIP_ERROR DeviceCommissioner::ProcessOpCSR(const ByteSpan & NOCSRElements, const ByteSpan & AttestationSignature)
+CHIP_ERROR DeviceCommissioner::SendTrustedRootCertificate(CommissioneeDeviceProxy * device, const ByteSpan & rcac)
```

Comment:
This lost the logging it used to have, no?

Reply:
I have added it back.